src/wallet/silentpayments.cpp

```diff
@@ -0,0 +1,294 @@
+                return true;
+            }
+            return false;
+        }), output_pub_keys.end());
+    } while (!output_pub_keys.empty() && keep_going);
+    return tweaks;
+}
+
+std::optional<std::pair<uint256, CPubKey>> GetSilentPaymentTweakDataFromTxInputs(const std::vector<CTxIn>& vin, const std::map<COutPoint, Coin>& coins)
+{
+    // Extract the keys from the inputs
+    // or skip if no valid inputs
+    std::vector<CPubKey> input_pubkeys;
+    std::vector<COutPoint> input_outpoints;
+    for (const CTxIn& txin : vin) {
+        const Coin& coin = coins.at(txin.prevout);
+        Assert(!coin.IsSpent());
+        input_outpoints.emplace_back(txin.prevout);
+
+        std::vector<std::vector<unsigned char>> solutions;
+        TxoutType type = Solver(coin.out.scriptPubKey, solutions);
+        if (type == TxoutType::WITNESS_V1_TAPROOT) {
+            // Check for H point in script path spend
+            if (txin.scriptWitness.stack.size() > 1) {
+                // Check for annex
+                bool has_annex = txin.scriptWitness.stack.back()[0] == ANNEX_TAG;
+                size_t post_annex_size = txin.scriptWitness.stack.size() - (has_annex ? 1 : 0);
+                if (post_annex_size > 1) {
+                    // Actually a script path spend
+                    const std::vector<unsigned char>& control = txin.scriptWitness.stack.at(post_annex_size - 1);
+                    Assert(control.size() >= 33);
+                    if (std::equal(NUMS_H.begin(), NUMS_H.end(), control.begin() + 1)) {
+                        // Skip script path with H internal key
+                        continue;
+                    }
+                }
+            }
+
+            std::vector<unsigned char> pubkey;
+            pubkey.resize(33);
+            pubkey[0] = 0x02;
+            std::copy(solutions[0].begin(), solutions[0].end(), pubkey.begin() + 1);
+            input_pubkeys.emplace_back(pubkey);
+        } else if (type == TxoutType::WITNESS_V0_KEYHASH) {
+            input_pubkeys.emplace_back(txin.scriptWitness.stack.back());
+        } else if (type == TxoutType::PUBKEYHASH || type == TxoutType::SCRIPTHASH) {
+            // Use the script interpreter to get the stack after executing the scriptSig
+            std::vector<std::vector<unsigned char>> stack;
+            ScriptError serror;
+            Assert(EvalScript(stack, txin.scriptSig, MANDATORY_SCRIPT_VERIFY_FLAGS, DUMMY_CHECKER, SigVersion::BASE, &serror));
+            if (type == TxoutType::PUBKEYHASH) {
+                input_pubkeys.emplace_back(stack.back());
+            } else if (type == TxoutType::SCRIPTHASH) {
+                // Check if the redeemScript is P2WPKH
+                CScript redeem_script{stack.back().begin(), stack.back().end()};
+                TxoutType rs_type = Solver(redeem_script, solutions);
+                if (rs_type == TxoutType::WITNESS_V0_KEYHASH) {
```

Comment:
@josibake p2sh-p2pkh or p2sh-p2pk are not addressed in the BIP. While I understand they are not commonly used scripts but `Solver(redeem_script, solutions)` can still detect p2pkh or p2pk in the redeem_script. Is there anything against adding support for these scripts? assuming that we ignore non-standard or malleated p2pkh in the redeem script

Reply:
The goal is not to support every possible script. Rather, we want to limit it to a sane subset to avoid technical debt and keep the implementation simple. In fact, the only reason for including `P2PKH` at all is due to the fact it is still widely used and makes up a significant percentage of the UTXO set, otherwise we would have only allowed Segwit script templates as this gives us non-malleability guarantees.

Reply:
Understood. Thanks for clearing that up @josibake